Signup Test Suite

meisekimiu opened the pull request:
This PR adds sign up, login, and cleanup test suites to the TypeScript functional test. This also creates the "Credentials" system, which saves logins between testing specs so that they can be reused between suites and cleaned up in the cleanup testing suite.

I tried to copy the algorithms used for generating usernames and passwords from the python test, so that is why the password generation code might look a bit weird (they were originally written in Python, but I decided to use the cryptographically secure random functions as well just to be safe).

Please leave any feedback on the `Credentials` object I created to handle created "sessions". I think eventually it should save both created *and* deleted sessions, so that suites know what the "auto increment" number should be when creating a new account after deletions, but I thought this was getting too into the weeds when that is an unusual (but not completely impossible) situation for running these tests. The module itself was designed under a lot of various compromises. Cypress tests run in a browser environment so Node.js specific code can't be run (such as `fs.readFileSync`) and in the `cypress.config.ts` file it's the opposite; the file is run in a Node.js specific domain so *browser* code won't run properly.

**Steps to test:**
1. `cd` into the `js` directory
2. Run `CYPRESS_BASE_URL=https://local.permanent.org npm run start`. You can change the BASE_URL to point to another environment.
3. Verify tests pass.

liam-lloyd commented on js/cypress/support/generatePassword.ts:
```diff
@@ -0,0 +1,13 @@
+
+export function generatePassword(length: number): string {
+  const lowestPrintableCode = 33;
+  const highestPrintableCode = 126 - lowestPrintableCode;
```
If I'm understanding correctly, this variable name is misleading, as 126 is the highest printable ASCII code, and this variable is the difference between the highest code and the lowest. If this was the highest printable code, the arithmetic at line 10 would produce non-printable codes.

liam-lloyd reviewed (commented):
The signup test looks good to me, but I'm concerned about the approach of using test suites to set up for or clean up after other test suites. I think this structure would eventually make these tests hard to maintain, compared to a structure where each test handles its own setup and teardown. Allowing the tests to depend on each other the way they do here couples the tests, which could eventually make it difficult to modify a test without understanding the whole repo. It also means we won't be able to run individual tests, or even test suites, without running the others, which could slow down debugging efforts.
Requiring each test to handle its own setup and teardown does potentially make the tests as a whole slower. To mitigate this, I'd suggest not using the UI to test things that aren't part of what a particular test tests. For example, a test for account deletion should proceed through the UI to delete an account, but the teardown for a signup test could just call the API. We could also consider adding some canonical test users to each environment so tests of logged-in functionality don't have to worry about creating or deleting users, although this requires ensuring that no one manually changes the state of those accounts.
Also, I think if we have support code that requires its own tests, as the credentials class does here, those tests should not run as part of the "functional tests", since they're really unit tests for this repo. It's possible this would be handled entirely by how we invoke the tests in CI though.

meisekimiu commented:
> We could also consider adding some canonical test users to each environment so tests of logged-in functionality don't have to worry about creating or deleting users, although this requires ensuring that no one manually changes the state of those accounts.

I'm a fan of this approach, since I would like to have these tests be independently runnable (something I struggled to pull off with this solution) and I don't want to be needlessly creating and destroying several users every single time we do a deployment. I think this data can be passed in via environment variables so each user and environment can have separate settings.

liam-lloyd commented:
> > We could also consider adding some canonical test users to each environment so tests of logged-in functionality don't have to worry about creating or deleting users, although this requires ensuring that no one manually changes the state of those accounts.
> 
> I'm a fan of this approach, since I would like to have these tests be independently runnable (something I struggled to pull off with this solution) and I don't want to be needlessly creating and destroying several users every single time we do a deployment. I think this data can be passed in via environment variables so each user and environment can have separate settings.
> 
> (This also frees up the dependency of signup from upload tests)

Yeah, I've seen some teams have trouble keeping people from messing with the test users, but I don't think that would be a problem for us.

meisekimiu commented:
I've adjusted the signup test suite code so it does cleanup within the suite (using the API directly), and deleted the unneeded code now that we'll be using preexisting accounts for other test suites.
